### PR TITLE
Discovery key fix

### PR DESCRIPTION
**Pull request by gmaclennan**
Re-work of #58. This adds tests, improves docs and throws instead of failing silently if an invalid projectKey is passed.

### CODE REVIEW

File: sync.js [hunk 1]
```diff
@@ -460,15 +459,23 @@ function WifiPeer (connection, info) {
...
+/**
+ * Generate a discovery key for a given projectKey. If projectKey is undefined
+ * then it will use SYNC_DEFAULT_KEY as the discovery key (this is for backwards
+ * compatibility with clients that did not use projectKeys)
+ *
+ * @param {String|Buffer} projectKey A unique random key identifying the
+ * project. Must be 32-byte Buffer or a string hex encoding of a 32-Byte buffer
+ */
+function discoveryKey (projectKey) {
+  if (typeof projectKey === 'undefined') return SYNC_DEFAULT_KEY
```
Comment (noffle) [inline sync.js]: Having a function that "generates a discovery key for a given projectKey" fall back to returning a completely unrelated value (`SYNC_DEFAULT_KEY`) feels weird. I think keeping it outside makes this function much simpler && clearer.

File: test/sync.js [hunk 2]
```diff
@@ -84,6 +85,77 @@ tape('sync: two servers find eachother', function (t) {
   })
 })
 
+tape('sync: two servers find each other with default same projectKey', function (t) {
```
Comment (noffle) [inline test/sync.js]:
```suggestion
tape('sync: two servers find each other with same projectKey', function (t) {
```

File: sync.js (hunk 1 above)
Comment (gmaclennan) [reply sync.js]: yeah, I get that. What do you think the best way of throwing is though? Test the key is valid in the join/leave functions and throw from there before trying to create a discovery key?

Comment (gmaclennan) [reply sync.js]: The alternative I can think of is to test the project key is valid in both the join and leave functions and throw there, but then this function is just crypto.discoveryKey and we've ended up copying the same code into two places.

Comment (noffle) [reply sync.js]: I'm ok with merging as-is and coming back to this later, which we can refactor /wo breaking any API stuff. :+1:

**Pull request merged**